### PR TITLE
fix race condition when creating new record for repo stat counter

### PR DESCRIPTION
**Pull request by xsuchy**
Previously we claimed that this method does commit. But it did just add() - and that does not commit. We do a commit now. And during a commit we check for integrity error. If that happens that means somebody inserted it just second ago. So lets try read again.

Fixes: #3073

### CODE REVIEW

**Comment by xsuchy** [comment]
Note that I did not tested this as I have no idea how to test it.

**Comment by FrostyX** [comment]
> Note that I did not tested this as I have no idea how to test it.

Same here

> Previously we claimed that this method does commit. But it did just add() - and that does not commit. We do a commit now.

In this case, I would say the code is as expected and the docstring is outdated. I am looking at the usages of the `incr` method and we always call `db.session.commit()` right after it. We do the same thing in unit tests. There is only one place that doesn't

https://github.com/fedora-copr/copr/blob/main/frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_rpmrepo.py#L146

isn't this our culprit?

**Comment by praiskup** [comment]
Per the initial commit c1a5794e39b3d1ea433cccc1c1cb55e52b72fd4a  it seems that the method was never "dirty" actually, and it never committed.

**Comment by praiskup** [comment]
> db.session.commit() right after it. We do the same thing in unit tests. There is only one place that doesn't
> isn't this our culprit?

This seems like the classic problem "get or create" race.  While incrementing the counter should be race-free, creating the object isn't.

I think we should separate the concerns of creating vs incrementing.  Ie. could we create the stats instance at the same time with the repository configuration (CoprChroot instance)?

**Comment by praiskup** [comment]
> incrementing the counter should be race-free

Hm, but not with ORM, I'm afraid.  We should do things like `update t set c = c + 1`.

**Comment by praiskup** [comment]
> I think we should separate the concerns of creating vs incrementing.

Namely, I don't think [this method](https://github.com/enricobarzetti/sqlalchemy_get_or_create/blob/master/sqlalchemy_get_or_create/__init__.py#L38) is worth it.  We just should have some form of `create_with_defaults()` method initializing the counter to 0 and doing the `commit()` (ignoring the possible IntegrityError).  And then atomically increment with `incr()`.

**Comment by praiskup** [comment]
See the [docs related to sessions](https://docs.sqlalchemy.org/en/20/orm/session_events.html), namely:

>  The reason the [Connection](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection) is passed is that it is encouraged that simple SQL operations take place here, directly on the [Connection](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection), such as incrementing counters or inserting extra rows within log tables.

**Comment by xsuchy** [comment]
Updated.
I removed the return() because the values were not used and therefore I could remove refresh().
It now use UPDATE statement.

**Comment by xsuchy** [comment]
Updated. I removed one filter() that was not needed as name is already primary index.

**Review by praiskup — APPROVED** [review]
LGTM

File: frontend/coprs_frontend/coprs/logic/stat_logic.py [hunk 1]
```diff
@@ -39,13 +40,17 @@ def incr(cls, name, counter_type, count=1):
         """
         try:
             csl = CounterStatLogic.get(name).one()
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/logic/stat_logic.py:42]: Sorry, I was confused with this.  This is an unused variable, but that's a nit.

**Pull request merged**